src/Utils.vala

```diff
@@ -427,45 +438,189 @@ public class FeedReader.Utils : GLib.Object {
+
+
+		// last chance: try allesedv service
+		string allesedv_url = siteURL;
+		if(allesedv_url.has_prefix("http://"))
+			allesedv_url = allesedv_url.replace("http://", "");
+		else if(allesedv_url.has_prefix("https://"))
+			allesedv_url = allesedv_url.replace("https://", "");
+		allesedv_url = "https://f1.allesedv.com/32/%s".printf(allesedv_url);
+
+		return downloadIcon(feed_id, allesedv_url, icon_path);
+	}
+
+	public static bool downloadIcon(string feed_id, string? icon_url, string icon_path = GLib.Environment.get_user_data_dir() + "/feedreader/data/feed_icons/")
+	{
+		if(icon_url == "" || icon_url == null || GLib.Uri.parse_scheme(icon_url) == null)
 		{
-			Logger.debug("Utils: downloadIcon() url = \"%s\"".printf(url));
+			Logger.warning(@"Utils.downloadIcon: icon_url not valid $icon_url");
+			return false;
+		}
 
-			Soup.Message message_dlIcon;
-			message_dlIcon = new Soup.Message ("GET", "https://f1.allesedv.com/32/%s".printf(url));
+		var path = GLib.File.new_for_path(icon_path);
+		try{path.make_directory_with_parents();}catch(GLib.Error e){}
+		string local_filename = icon_path + feed_id.replace("/", "_").replace(".", "_") + ".ico";
+		string etag_filename = icon_path + feed_id.replace("/", "_").replace(".", "_") + ".txt";
 
+		// file already exists
+		if(FileUtils.test(local_filename, GLib.FileTest.EXISTS)
+		|| !FileUtils.test(etag_filename, GLib.FileTest.EXISTS))
+		{
+			var message_head = new Soup.Message("HEAD", icon_url);
 			if(Settings.tweaks().get_boolean("do-not-track"))
-				message_dlIcon.request_headers.append("DNT", "1");
+				message_head.request_headers.append("DNT", "1");
 
-			var status = session.send_message(message_dlIcon);
-			if (status == 200)
+			var status = getSession().send_message(message_head);
+			if(status == 200)
 			{
-				try
+				string etag = message_head.response_headers.get_one("ETag");
+
+				if(etag != null)
 				{
-					FileUtils.set_contents(local_filename, (string)message_dlIcon.response_body.flatten().data, (long)message_dlIcon.response_body.length);
+					if(FileUtils.test(etag_filename, GLib.FileTest.EXISTS)
+					&& etag == getFileContent(etag_filename))
+					{
+						// file exists and is identical to remote file
+						Logger.debug(@"Utils.downloadIcon: etag identical -> icon unchanged $feed_id");
+						return true;
+					}
+					else
+					{
+						Logger.debug(@"Utils.downloadIcon: write etag $etag");
+						try
+						{
+							FileUtils.set_contents(etag_filename, etag);
+						}
+						catch(GLib.FileError e)
+						{
+							Logger.error("Error writing etag: %s".printf(e.message));
+						}
+					}
 				}
-				catch(GLib.FileError e)
+				else
 				{
-					Logger.error("Error writing icon: %s".printf(e.message));
+					Logger.warning(@"no etag for $icon_url");
 				}
-				return true;
 			}
-			Logger.error(@"Error downloading icon for feed: $feed_id");
-			return false;
+			else
+			{
+				Logger.warning(@"HEAD-request failed $icon_url");
+				return false;
+			}
```

Comment:
Rather than doing a HEAD and checking for etag, we should send the etag with the GET request and check for a 304 response.